### PR TITLE
sample multinomial without replacement - GPU

### PR DESCRIPTION
**Pull request by aalmah**
This is a first working version. It's not very efficient, so we need to improve on that. This is for #4139

### CODE REVIEW

File: theano/gpuarray/multinomial.py [hunk 1]
```diff
@@ -226,6 +228,244 @@ def c_code(self, node, name, inp, outputs, sub):
...
+
+    def gpu_kernels(self, node, name):
+        code = """
+KERNEL void k_multi_warp_multinomial_wor(
+    const ga_size nb_multi,
+    const ga_size nb_outcomes,
+    const ga_size n_samples,
+    // GLOBAL_MEM int * ga_n_samples,
+    GLOBAL_MEM float * global_pvals_copy,
+    // GLOBAL_MEM float * global_pvals,
```
Comment (abergeron) [inline theano/gpuarray/multinomial.py]: I wouldn't keep these comments.

File: theano/gpuarray/multinomial.py [hunk 2]
```diff
@@ -226,6 +228,244 @@ def c_code(self, node, name, inp, outputs, sub):
...
+    dims[1] = PyGpuArray_DIMS(pvals)[0];
+
+    if (theano_prep_output(&out, 2, dims, GA_LONG,
+                           GA_C_ORDER, %(ctx)s) != 0){
+      %(fail)s
+    }
+    
+    %(out)s = out;
+        
+    GpuArray_memset(&(out->ga), -1);
```
Comment (abergeron) [inline theano/gpuarray/multinomial.py]: You should probably remove this for the final code.

File: theano/gpuarray/multinomial.py [hunk 3]
```diff
@@ -226,6 +228,244 @@ def c_code(self, node, name, inp, outputs, sub):
...
+        int nb_multi = PyGpuArray_DIMS(pvals)[0];
+        int nb_outcomes = PyGpuArray_DIMS(pvals)[1];
+        //TODO : change this for a beautiful constant
+        int max_nb_blocks = 2<<15 - 1;
+        size_t nb_blocks = max_nb_blocks + 1;
+        size_t nb_threads=16; // so it really starts at 32, because of the *2
+        do
+        {
+            nb_threads*=2;
+            if (nb_multi % %nb_threads == 0)
```
Comment (abergeron) [inline theano/gpuarray/multinomial.py]: Maybe fix this confusing syntax to ` %% `

File: theano/gpuarray/multinomial.py [hunk 4]
```diff
@@ -226,6 +228,244 @@ def c_code(self, node, name, inp, outputs, sub):
...
+                nb_blocks = nb_multi/nb_threads;
+            else
+                nb_blocks = (int)((float)nb_multi/(float)nb_threads + 1.);
+        } while (nb_blocks > max_nb_blocks);
+
+        //printf("\\nN=%%i b=%%i t=%%i t*b=%%i",
+        //         nb_multi, nb_blocks, nb_threads, nb_blocks*nb_threads);
```
Comment (abergeron) [inline theano/gpuarray/multinomial.py]: Remove this comment.

File: theano/gpuarray/multinomial.py [hunk 5]
```diff
@@ -226,6 +228,244 @@ def c_code(self, node, name, inp, outputs, sub):
...
+            %(fail)s;
+        }
+        if(%(sync)d)
+            GpuArray_sync(&(out->ga));
+    } // END NESTED SCOPE
+        """ % locals()
+        return s
+
+    def c_code_cache_version(self):
+        return None
```
Comment (abergeron) [inline theano/gpuarray/multinomial.py]: Define a real version for the cache.

File: theano/gpuarray/multinomial.py [hunk 6]
```diff
@@ -247,3 +487,19 @@ def local_gpua_multinomial(node, context_name):
...
+    p, u, n = node.inputs
+    # try:
+    #     if get_scalar_constant_value(n_samples) != 1:
+    #         return None
+    # except NotScalarConstantError:
+    #     return None
+    m, = node.outputs
+    if ((p.dtype == u.dtype == 'float32') and (m.dtype == 'int64')):
+        gpu_op = GPUAMultinomialWOReplacementFromUniform(node.op.odtype)
+        return gpuarray.elemwise.GpuDimShuffle([False, False], [1, 0])(
```
Comment (abergeron) [inline theano/gpuarray/multinomial.py]: Why is the output transposed?

File: theano/gpuarray/tests/test_multinomial_wo_replacement.py [hunk 7]
```diff
@@ -0,0 +1,196 @@
+from __future__ import absolute_import, print_function, division
```
Comment (abergeron) [inline theano/gpuarray/tests/test_multinomial_wo_replacement.py]: Name this file test_multinomial.py

File: theano/gpuarray/tests/test_multinomial_wo_replacement.py [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+from theano.sandbox.rng_mrg import MRG_RandomStreams as RandomStreams
+import unittest
+from .config import mode_with_gpu, mode_without_gpu
+from ..multinomial import GPUAMultinomialWOReplacementFromUniform
+
+def get_mode(gpu):
+    mode = mode_without_gpu
+    if gpu:
+        mode = mode_with_gpu
+    return mode
```
Comment (abergeron) [inline theano/gpuarray/tests/test_multinomial_wo_replacement.py]: This function doesn't seem especially useful.

**Comment by abergeron** [comment]
I've made a couple of comments, but otherwise it's ok.

File: theano/gpuarray/multinomial.py (hunk 6 above)
Comment (aalmah) [inline theano/gpuarray/multinomial.py]: This is something I picked up from the with-replacement code. My understanding is that it makes memory fetching more efficient.

File: theano/gpuarray/tests/test_multinomial_wo_replacement.py (hunk 7 above)
Comment (aalmah) [inline theano/gpuarray/tests/test_multinomial_wo_replacement.py]: There's already a file with this name. Do you want to merge the two files?

Comment (nouiz) [inline theano/gpuarray/tests/test_multinomial_wo_replacement.py]: yes.

File: theano/gpuarray/multinomial.py (hunk 6 above)
Comment (nouiz) [inline theano/gpuarray/multinomial.py]: Can you quickly check if this is still the case with this code?

**Comment by abergeron** [comment]
There are still two flake8 problems:
```
/home/travis/build/Theano/Theano/theano/gpuarray/multinomial.py:486:1: E302 expected 2 blank lines, found 1
/home/travis/build/Theano/Theano/theano/gpuarray/tests/test_multinomial_wo_replacement.py:177:36: E226 missing whitespace around arithmetic operator
```

Also it appears the branch has conflicts with the current master and will need a rebase.

**Comment by abergeron** [comment]
Finished in #4757

**Pull request closed without merging**